share/smack/lib/pthread.c

```diff
@@ -0,0 +1,253 @@
+#include "pthread.h"
```

Comment:
I don’t think header files (`.h` files) should be considered as specifications of what *resources* implementation files (`.cpp` files) use. I think header files should be considered as *interface* specifications: they should specify what the implementations of that interface provide. The implementation files themselves ought to include any additional required resources.

Consider a module `B.cpp` which includes `A.h` whose methods are implemented in `A.cpp`. The `B.cpp` module needs to know about how it can interact with `A`. It does not need to know how `A.cpp` was implemented, e.g., whether it used the C library `time.h`. This is an implementation detail, that has no business being mentioned in `A.h` — unless, of course, knowledge of the interface provided by `time.h` is required for using the interface provided by `A.h`, e.g., that some of `A`’s methods have parameters with types provided by `time.h`.

Besides the conceptual idea of modularity, this difference also impacts compile-time performance: including the header files used by other modules (e.g., by `A.cpp`) when compiling any given module (e.g., `B.cpp`) increases the total number of headers included for each module, and results in slower compilation time.

It is true that we have been abusing this all along while developing SMACK. I think that was a mistake which should be corrected as we go along.

I do not know where your convention of ensuring that each implementation includes a single header file comes from, but I think it is a bad convention, and one that is certainly not followed by LLVM. Just look at any LLVM implementation file, e.g,.

  http://llvm.org/docs/doxygen/html/Function_8cpp_source.html

Further, my points are directly mentioned in the LLVM coding standards:

http://llvm.org/docs/CodingStandards.html#a-public-header-file-is-a-module
http://llvm.org/docs/CodingStandards.html#include-as-little-as-possible

Touchée! ;-)